src/coap_prng.c

```diff
@@ -55,7 +55,8 @@ coap_prng_impl( unsigned char *buf, size_t len ) {
 static int
 coap_prng_default(void *buf, size_t len) {
 #if defined(MBEDTLS_ENTROPY_HARDWARE_ALT)
-  return mbedtls_hardware_poll(NULL, buf, len, NULL);
+  /* mbedtls_hardware_poll() returns 0 on success */
+  return (mbedtls_hardware_poll(NULL, buf, len, NULL) ? 0 : 1);
 #else /* !MBEDTLS_ENTROPY_HARDWARE_ALT */
 #ifdef HAVE_GETRANDOM
   return getrandom(buf, len, 0);
```

Comment:
[getrandom](https://man7.org/linux/man-pages/man2/getrandom.2.html) may require also some handling.

Reply:
Good catch.  Fixed `coap_rng()` to now test for positive return from `coap_prng()` so that future code callers of `coap_prng()` can test for `getrandom()` errors.

Reply:
Starting with the "root-cause" obstacle, that a function (of an other project) has no definition of the expected return value, I would prefer to have a clear one here:

- 1 := success, 0 := failure
- len := success, -1 := failure

I don't think, a mix is a too good idea in midterm, it only helps for short term.

Reply:
mbedtls_*() function coding standards for function returns appear to be -ve on error, 0 OK and rarely +ve OK.  See https://tls.mbed.org/kb/development/mbedtls-coding-standards section "Return type". `mbedtls_hardware_poll()` would fall into the category of -ve fail, 0 OK (as does `f_rng `for that matter).

Given that coap_rng() now handles 0 or negative returns from coap_prng() as failure, I believe that this code is OK, but can make the failure return -1 for the `mbedtls_hardware_poll()` case (but then should we be setting `errno`?).

Reply:
I was addressing the return value of `coap_prng_default`, which seems to be somehow ambiguous to me.

Reply:
Understood.  I have taken to documenting the return values for some functions at the start of the code where it is not documented in a header file for clarity to the reader.

So coap_prng() only ever returns 0 on failure and 1 on success (as documented in the header file).  For non LwIP and non Contiki systems, `coap_prng()` calls `coap_prng_default()` (or user supplied function) but never checks the return value for an error (which should be used to trigger a 0 return).

[LwIP and Contiki appear to only ever return 1 for `coap_prng()`.]

I guess therefore that `coap_prng_default()` (or user supplied function) should return 0 for failure and 1 for success so that coap_prng() can simply return `rand_func`'s return.

Reply:
Thanks!

Reply:
I have just pushed new code so that `coap_prng_default() `is aligned with return values as defined for the `coap_rand_func_t` definition (1 success, 0 failure) which can be used as an alternative to `coap_prng_default()` if `coap_set_prng()` is invoked.